cluster-api/pkg/controller/machineset/controller.go

```diff
@@ -0,0 +1,218 @@
+			}
+		},
+		UpdateFunc: func(old interface{}, new interface{}) {
+			key, err := cache.MetaNamespaceKeyFunc(new)
+			if err == nil {
+				controller.queue.AddRateLimited(key)
+			}
+		},
+	})
+
+	return controller
+}
+
+// Run starts the control loop. This method is blocking.
+func (c *Controller) Run(threadiness int, stopCh <-chan struct{}) {
+	defer runtime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	glog.Info("starting machineset controller")
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(c.runWorker, time.Second, stopCh)
+	}
+
+	<-stopCh
+	glog.Info("shutting down machineset controller")
+}
+
+func (c *Controller) runWorker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+// processNextWorkItem deals with one key off the queue.  It returns false
+// when it's time to quit.
+func (c *Controller) processNextWorkItem() bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+
+	err := c.syncHandler(key.(string))
+	if err == nil {
+		c.queue.Forget(key)
+		return true
+	}
+
+	runtime.HandleError(fmt.Errorf("%v failed with: %v", key, err))
+	c.queue.AddRateLimited(key)
+
+	return true
+}
+
+// syncHandler holds the controller's business logic.
+// it makes sure that the current state is equal to the desired state.
+// note that the current state of the cluster is calculated based on the number of machines
+// that are owned by the given machineSet (key).
+func (c *Controller) syncHandler(key string) error {
+	machineSet, err := c.machineSetsLister.Get(key)
+	if err != nil {
+		return err
+	}
+
+	filteredMachines, err := c.getMachines(machineSet)
+	if err != nil {
+		return err
+	}
+
+	return c.syncReplicas(machineSet, filteredMachines)
+}
+
+// syncReplicas essentially scales machine resources up and down.
+func (c *Controller) syncReplicas(machineSet *v1alpha1.MachineSet, machines []*v1alpha1.Machine) error {
+	currentMachineCount := int32(len(machines))
+	desiredReplicas := int32(0)
+	if machineSet.Spec.Replicas != nil {
+	 desiredReplicas = *machineSet.Spec.Replicas
+	}
+
+	if desiredReplicas > currentMachineCount {
+		glog.V(2).Infof("creating a machine ( spec.replicas(%d) > currentMachineCount(%d) )", desiredReplicas, currentMachineCount)
+		machine, err := c.createMachine(machineSet)
+		if err != nil {
+			return err
+		}
+		_, err = c.machineClient.ClusterV1alpha1().Machines().Create(machine)
+		if err != nil {
+			return err
+		}
+		if err = waitForMachineInCache(machine.Name, c.machineLister); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	if desiredReplicas < currentMachineCount {
+		glog.V(2).Infof("deleting a machine ( spec.replicas(%d) < currentMachineCount(%d) )", desiredReplicas, currentMachineCount)
+		machineToDelete := machines[0].Name
+		err := c.machineClient.ClusterV1alpha1().Machines().Delete(machineToDelete, &metav1.DeleteOptions{})
+		if err != nil {
+			return err
+		}
+		if err = waitForMachineDeletedInCache(machineToDelete, c.machineLister); err != nil {
```

Comment:
Why is this needed? You should only begin the reconcile loop when **all** your caches are populated.